Workaround missing G4persistency target bug in Geant4 11.2

drbenmorgan opened the pull request:
As reported on Slack, and the same report on [VecGeom JIRA](https://its.cern.ch/jira/browse/VECGEOM-621), Geant4 11.2 split the `G4persistency` target into subcomponents:

- `G4mctruth`
- `G4geomtext`
- `G4gdml` (optional, only if Geant4 built with GDML support)

and so projects use the direct CMake targets will fail to link with these versions. A bug fix will be made in Geant4 itself to address this, but we'll still need to support the affected versions of 11.2.

Workaround this issue in Celeritas by translating request for `persistency` target in `celeritas_get_g4libs` to appropriate underlying targets. Translation rather than creation of our own imported target was chosen to:

- Avoid possible (albeit unlikely) clashes with other projects that might create the same target (and likewise not mess with the `Geant4::` CMake namespace).
- Reduce dependence on use of Celeritas' FindGeant4 wrapper by downstream projects.

drbenmorgan committed:
Workaround missing G4persistency target bug in Geant4 11.2

Geant4 11.2 split the G4persistency target into subcomponents:

- G4mctruth
- G4geomtext
- G4gdml (optional, only if Geant4 built with GDML support)

and so projects use the direct CMake targets will fail to link
these versions. A bug fix will be made in Geant4 itself to address
this.

Workaround issue in Celeritas by translating request for "persistency"
target in celeritas_get_g4libs to appropriate underlying targets. Use
translation rather than creation of our own imported target to:

- avoid possible (albeit unlikely) clashes with other projects that
  might create the same target.
- Reduce dependence on use of Celeritas' FindGeant4 wrapper by downstream
  projects.

pcanal committed:
Ignore request for G4tasking if G4 11.2+

drbenmorgan commented on cmake/CeleritasUtils.cmake:
```diff
@@ -647,6 +647,10 @@ function(celeritas_get_g4libs var)
     foreach(_lib IN LISTS ARGN)
       set(_lib Geant4::G4${_lib}${_ext})
 
+      # Tasking was merged into core management
+      if(_lib MATCHES "tasking" AND NOT TARGET "${_lib}")
+        continue()
+      endif()
```
If we call

```
celeritas_get_g4libs(OUTPUT_VAR tasking)
```

then `OUTPUT_VAR` would be empty... The problem here is if `tasking` is requested, then do we want the stuff in `G4global`, or that in `G4run`? I'm guessing that in _most_ cases, it's going to be `G4run`, but I'd have to look at the specifics of each call.

sethrj replied:
Maybe we just silently omit libraries that aren't there, and specify `tasking run` as arguments so that one or the other will get picked up?

drbenmorgan replied:
Yes, that would work, we'll just have to be aware of that should those calls change.

sethrj replied:
Well, how about we do it with the manual library checks and matching, and use the newer (more granular) libraries in the *calls* to this function and map them to the older (coarser) `persistency` library.

drbenmorgan replied:
I think for now it's probably better to stick with the existing names _for now_. We could review this depending on what is done in Geant4 11.2.2/11.3 with bug fixes in the former and what granularity is arrived at in 11.3.

sethrj commented on cmake/CeleritasUtils.cmake:
```diff
@@ -646,10 +646,27 @@ function(celeritas_get_g4libs var)
+        endif()
       else()
-        list(APPEND _result "${_lib}")
+        if(NOT TARGET "${_lib}")
```
The cmake would probably be clearer if this were all inverted (and since we're reusing the "short" library name input by the user, it would make it easier not to rewrite it on the first line:
```cmake
foreach(_shortlib IN LISTS ARGN)
set(_lib Geant4::G4${_shortlib}${_ext})
if(TARGET "${_lib}")
  list(APPEND _result "${_lib}")
elseif(_short_lib STREQUAL "persistency")
  ....
elseif(...)
...
endif()
```

drbenmorgan replied:
Yes, that's much nicer - implemented in the latest commit.

drbenmorgan committed:
Simplify logic of Geant4 target checking

Check obvious case first, then the special cases for persistency
and tasking. Post-process list to remove duplicates to avoid
ld warnings observed on macOS.

sethrj reviewed (approved):
Nice! Thanks for that.

sethrj committed:
Merge branch 'develop' into fix-g4persistency-target
